src/Features/Core/Portable/Testing/AbstractTestMethodFinder.cs

```diff
@@ -0,0 +1,92 @@
+
+    protected readonly ImmutableArray<ITestFrameworkMetadata> TestFrameworkMetadata = testFrameworks.ToImmutableArray();
+
+    protected abstract Task<bool> IsTestMethodAsync(Document document, TMethodDeclaration method, CancellationToken cancellationToken);
+
+    public async Task<ImmutableArray<SyntaxNode>> GetPotentialTestMethodsAsync(TextSpan textSpan, Document document, CancellationToken cancellationToken)
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        var testNodes = await GetPotentialTestMethodsAsync(document, cancellationToken).ConfigureAwait(false);
+
+        // Find any test methods that intersect with the requested span.
+        var intersectingNodes = testNodes.Where(node => node.Span.IntersectsWith(textSpan)).ToImmutableArray();
+        if (!intersectingNodes.IsEmpty)
+        {
+            return intersectingNodes;
+        }
+
+        // We might have been invoked on a test class.  Check if any of the test method parent nodes intersect with the requested text span.
+        return testNodes.Where(node => node.Parent?.Span.IntersectsWith(textSpan) == true).ToImmutableArray();
+    }
+
+    public async Task<bool> IsMatchAsync(SyntaxNode node, string fullyQualifiedTestName, Document document, CancellationToken cancellationToken)
+    {
+        var method = node as TMethodDeclaration;
+        Contract.ThrowIfNull(method, $"Node with kind {node.RawKind} is not a method");
+
+        // Since discovered tests are not guarantied to run on a particular snapshot, we match optimistically based on test name.
+        var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        var methodSymbol = semanticModel.GetDeclaredSymbol(method, cancellationToken);
+        Contract.ThrowIfNull(methodSymbol, "Test method has no symbol");
+
+        var fullyQualifiedMethodName = methodSymbol.ToDisplayString(s_methodSymbolNoParametersDisplayFormat);
+        var fullyQualifiedMethodNameWithParameters = methodSymbol.ToDisplayString(s_methodSymbolWithParametersDisplayFormat);
+        return fullyQualifiedMethodName == fullyQualifiedTestName || fullyQualifiedMethodNameWithParameters == fullyQualifiedTestName;
+    }
+
+    public Task<bool> IsTestMethodAsync(Document document, SyntaxNode node, CancellationToken cancellationToken)
+    {
+        return node is TMethodDeclaration method ? IsTestMethodAsync(document, method, cancellationToken) : SpecializedTasks.False;
+    }
+
+    private async Task<ImmutableArray<SyntaxNode>> GetPotentialTestMethodsAsync(Document document, CancellationToken cancellationToken)
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        var methodsInRange = root.DescendantNodesAndSelf(descendIntoTrivia: false).OfType<TMethodDeclaration>();
```

Comment:
this is also surprisingly expensive (i just fixed up an analyzer that was doing this).  Consider a dedicated walk over namespaces, types, then methods.  It's not difficult to write up.  We could perhaps extract an existing walk into a helper as well if that would be beneficial.